docs/tutorials/time-series.md

```diff
@@ -0,0 +1,738 @@
+The fifth row violates lateness, since it is 100 minutes late with
+respect to the fourth row, and will be discarded.
+
+## How Feldera garbage collects old state
+
+Consider the `purchase` table from before.  As new purchase records are added
+over time, the table will keep growing.  How much of its history do we need to
+store?  A conventional database, serving as a system of record, must store every
+record added to it, until it is deleted by the user, so that it can answer
+queries over this data on-demand.  In contrast, the Feldera query engine
+incrementally maintains a set of views defined ahead of time and
+only keeps state needed by those views.
+
+:::note
+
+Feldera also supports on-demand queries (also know as **ad hoc queries**) for
+materialized tables and views.  See [documentation](/sql/materialized) for details.
+
+:::
+
+Let’s assume the `purchase` table is used to compute a single view that tracks
+the daily maximum purchase amount:
+
+```sql
+CREATE VIEW daily_max AS
+SELECT
+    TIMESTAMP_TRUNC(ts, DAY) as d,
+    MAX(amount) AS max_amount
+FROM
+    purchase
+GROUP BY
+    TIMESTAMP_TRUNC(ts, DAY);
+```
+
+If the `purchase` table did not have the `LATENESS` annotation,
+the query engine would assume that any record in this table can be inserted or
+deleted at any time.  Deleting an old record requires re-computing
+the daily maximum for the corresponding date.  This in turn requires storing
+the entire purchase history permanently.
+
+**We can do better with `LATENESS`**.  The `LATENESS` annotation guarantees
+that changes to the table cannot go more than one hour back.  Hence, daily
+maxima can only change for the current and, possibly, previous day.  Since
+we will never need to reevaluate the query for earlier dates, we do not need
+to store older data.
+
+To make this more precise, we need two new concepts:
+
+1. **Stream time** of a relation (a table or view) with respect to a timestamp column
+   is the highest timestamp value of any record that has ever been added to the relation.
+   Stream time can only move forward; removing the latest timestamp from the relation
+   does not cause it to revert. This concept enables reasoning about recent vs old
+   records without depending on a physical clock.
+
+2. **Waterline** of a relation with respect to a timestamp column is a timestamp value
+   `W`, such that all future additions or deletions in the relation will have timestamps
+   greater than or equal to `W`.
+
+### The data retention algorithm
+
+Feldera uses the following (simplified) procedure to identify and discard unused
+data:
+
+1. **Compute waterlines of all program relations** through a combination of offline
+   and online dataflow analysis:
+
+   * Given a SQL table with a timestamp column with the current stream time
+     `T` and lateness `L`, waterline `W` for this column is computed as: `W = T - L`.
+
+   * The waterline of a view column is derived from the waterlines
```

Comment:
I had this discussion about using waterlines to launch a denial of service attack by having records very far in the future. that deserves some mention. Or did I put that in the blog post?

Reply:
I think it's in the blog. I'll add a reference to that as further reading